solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol

```diff
@@ -0,0 +1,131 @@
+
+    // allows execution by the owner only
+    modifier ownerOnly {
+        require(hasRole(ROLE_OWNER, msg.sender), "ERR_ACCESS_DENIED");
+        _;
+    }
+
+    // allows execution by the seeder only
+    modifier seederOnly {
+        require(hasRole(ROLE_SEEDER, msg.sender), "ERR_ACCESS_DENIED");
+        _;
+    }
+
+    constructor() public {
+        // set up administrative roles.
+        _setRoleAdmin(ROLE_OWNER, ROLE_OWNER);
+        _setRoleAdmin(ROLE_SEEDER, ROLE_OWNER);
+
+        // allow the deployer to initially govern the contract.
+        _setupRole(ROLE_OWNER, msg.sender);
+    }
+
+    /**
+     * @dev increases the total amounts
+     *
+     * @param _provider         liquidity provider address
+     * @param _poolToken        pool token address
+     * @param _reserveToken     reserve token address
+     * @param _poolAmount       pool token amount
+     * @param _reserveAmount    reserve token amount
+     */
+    function increaseTotalAmounts(
+        address _provider,
+        IDSToken _poolToken,
+        IERC20Token _reserveToken,
+        uint256 _poolAmount,
+        uint256 _reserveAmount
+    ) external override ownerOnly {
+        totalPoolAmount[_poolToken] = totalPoolAmount[_poolToken].add(_poolAmount);
+        totalReserveAmount[_poolToken][_reserveToken] = totalReserveAmount[_poolToken][_reserveToken].add(_reserveAmount);
+        totalProviderAmount[_poolToken][_reserveToken][_provider] = totalProviderAmount[_poolToken][_reserveToken][_provider].add(_reserveAmount);
+    }
+
+    /**
+     * @dev decreases the total amounts
+     *
+     * @param _provider         liquidity provider address
+     * @param _poolToken        pool token address
+     * @param _reserveToken     reserve token address
+     * @param _poolAmount       pool token amount
+     * @param _reserveAmount    reserve token amount
+     */
+    function decreaseTotalAmounts(
+        address _provider,
+        IDSToken _poolToken,
+        IERC20Token _reserveToken,
+        uint256 _poolAmount,
+        uint256 _reserveAmount
+    ) external override ownerOnly {
+        totalPoolAmount[_poolToken] = totalPoolAmount[_poolToken].sub(_poolAmount);
+        totalReserveAmount[_poolToken][_reserveToken] = totalReserveAmount[_poolToken][_reserveToken].sub(_reserveAmount);
+        totalProviderAmount[_poolToken][_reserveToken][_provider] = totalProviderAmount[_poolToken][_reserveToken][_provider].sub(_reserveAmount);
+    }
+
+    /**
+     * @dev resets the total amounts
+     *
+     * @param _provider         liquidity provider address
+     * @param _poolToken        pool token address
+     * @param _reserveToken     reserve token address
+     */
+    function resetTotalAmounts(
+        address _provider,
+        IDSToken _poolToken,
+        IERC20Token _reserveToken
+    ) external override ownerOnly {
+        totalPoolAmount[_poolToken] = 0;
+        totalReserveAmount[_poolToken][_reserveToken] = 0;
```

Comment:
This looks like a bug. You shouldn't set total amounts to 0 if a single provider removes his liquidity. Do you have a test for this?

Reply:
@lbeder : Fixed; can you please check?